### PR TITLE
Fix (Series|DataFrame).interpolate for datetime dtypes

### LINKED ISSUE #19199: Interpolating of datetime64 values ignored (at best)
tz naive case:
```
index = pd.Index([23, 26, 30])
dti = pd.DatetimeIndex(['2015-09-23', '2015-09-26', '2015-09-30'])
ser = pd.Series(dti, index=index).reindex(range(23, 31))
df = pd.DataFrame(ser)

>>> ser    # for reference
23   2015-09-23
24          NaT
25          NaT
26   2015-09-26
27          NaT
28          NaT
29          NaT
30   2015-09-30
dtype: datetime64[ns]

>>> ser.interpolate()
23   2015-09-23
24          NaT
25          NaT
26   2015-09-26
27          NaT
28          NaT
29          NaT
30   2015-09-30
dtype: datetime64[ns]

>>> df.interpolate()
            0
23 2015-09-23
24        NaT
25        NaT
26 2015-09-26
27        NaT
28        NaT
29        NaT
30 2015-09-30
```

tz-aware case
```
index = pd.Index([23, 26, 30])
dti = pd.DatetimeIndex(['2015-09-23', '2015-09-26', '2015-09-30'], tz='US/Central')
ser = pd.Series(dti, index=index).reindex(range(23, 31))
df = pd.DataFrame(ser)

>>> ser
23   2015-09-23 00:00:00-05:00
24                         NaT
25                         NaT
26   2015-09-26 00:00:00-05:00
27                         NaT
28                         NaT
29                         NaT
30   2015-09-30 00:00:00-05:00
dtype: datetime64[ns, US/Central]

>>> ser.interpolate()
23   2015-09-23 00:00:00-05:00
24                         NaT
25                         NaT
26   2015-09-26 00:00:00-05:00
27                         NaT
28                         NaT
29                         NaT
30   2015-09-30 00:00:00-05:00
dtype: datetime64[ns, US/Central]

>>> df.interpolate()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/site-packages/pandas/core/generic.py", line 4738, in interpolate
    'object') == len(_maybe_transposed_self.T):
  File "/usr/local/lib/python2.7/site-packages/pandas/core/frame.py", line 1909, in transpose
    return super(DataFrame, self).transpose(1, 0, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/pandas/core/generic.py", line 599, in transpose
    new_values = self.values.transpose(axes_numbers)
  File "/usr/local/lib/python2.7/site-packages/pandas/core/base.py", line 696, in transpose
    nv.validate_transpose(args, kwargs)
  File "/usr/local/lib/python2.7/site-packages/pandas/compat/numpy/function.py", line 54, in __call__
    self.defaults)
  File "/usr/local/lib/python2.7/site-packages/pandas/util/_validators.py", line 218, in validate_args_and_kwargs
    validate_kwargs(fname, kwargs, compat_args)
  File "/usr/local/lib/python2.7/site-packages/pandas/util/_validators.py", line 157, in validate_kwargs
    _check_for_default_values(fname, kwds, compat_args)
  File "/usr/local/lib/python2.7/site-packages/pandas/util/_validators.py", line 69, in _check_for_default_values
    format(fname=fname, arg=key)))
ValueError: the 'axes' parameter is not supported in the pandas implementation of transpose()
```

The ValueError for the DataFrame case is in issue #19198.

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #19199
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/internals.py [hunk 1]
```diff
@@ -2592,6 +2577,44 @@ def set(self, locs, values, check=False):
...
+        if self.is_datetimetz:
+            # Convert to UTC for interpolation
+            data = data.tz_convert('UTC').values
+        if self.ndim > 1:
+            # DataFrame
+            data = np.atleast_2d(data)
+            mask = np.atleast_2d(mask)
+        data = data.astype(np.float64)
+        data[mask] = np.nan
+
```
Comment (jreback) [inline pandas/core/internals.py]: ths is a hodgepode. use the hierarchy to make this code readable. IOW put the datetime stuff in the dateimeblock.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/internals.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/internals.py]: This is in the DatetimeBlock.  I'm confused.

Comment (jreback) [reply pandas/core/internals.py]: then why would you need is_datetime.

Comment (jreback) [reply pandas/core/internals.py]: this needs to be be way less if/then stuff.

Comment (jbrockmendel) [reply pandas/core/internals.py]: > then why would you need is_datetime.

It's specifically `is_datetimetz`; the tzaware case gets slightly different handling.  If you're suggesting a separate implementation for DatetimeTZBlock vs DatetimeBlock I guess that's OK, but pretty sure there'll be complaints about code duplication that way.

File: pandas/core/generic.py [hunk 2]
```diff
@@ -5151,8 +5151,11 @@ def interpolate(self, method='linear', axis=0, limit=None, inplace=False,
             raise ValueError("Only `method=linear` interpolation is supported "
                              "on MultiIndexes.")
 
-        if _maybe_transposed_self._data.get_dtype_counts().get(
-                'object') == len(_maybe_transposed_self.T):
+        dtype_counts = _maybe_transposed_self._data.get_dtype_counts()
+        if ('object' in dtype_counts and
+                dtype_counts.get('object') == len(_maybe_transposed_self.T)):
+            # Checking for 'object' lets us avoid sometimes-fragile tranpose
```
Comment (jreback) [inline pandas/core/generic.py]: huh? where are you testing this

**Review by jreback — CHANGES_REQUESTED** [review]
you basically tripled the code here. pls don't do that. refactor into smaller more logical units.

**Comment by jbrockmendel** [comment]
> you basically tripled the code here. pls don't do that. refactor into smaller more logical units.

You complained about the existence of `if/else` conditions in the previous version when `DatetimeBlock` handled both the aware and unaware cases.  Are you suggesting going back to that or something else entirely?

File: pandas/core/generic.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/generic.py]: This is a workaround until #19198 is fixed:

```
dti = pd.date_range('2016-01-01', periods=3, tz='US/Pacific').insert(1, pd.NaT)
ser = pd.Series(dti)
df = ser.to_frame()

>>> df.interpolate()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pandas/core/generic.py", line 5155, in interpolate
    'object') == len(_maybe_transposed_self.T):
  File "pandas/core/frame.py", line 1941, in transpose
    return super(DataFrame, self).transpose(1, 0, **kwargs)
  File "pandas/core/generic.py", line 616, in transpose
    new_values = self.values.transpose(axes_numbers)
  File "pandas/core/base.py", line 701, in transpose
    nv.validate_transpose(args, kwargs)
  File "pandas/compat/numpy/function.py", line 54, in __call__
    self.defaults)
  File "pandas/util/_validators.py", line 218, in validate_args_and_kwargs
    validate_kwargs(fname, kwargs, compat_args)
  File "pandas/util/_validators.py", line 157, in validate_kwargs
    _check_for_default_values(fname, kwds, compat_args)
  File "pandas/util/_validators.py", line 69, in _check_for_default_values
    format(fname=fname, arg=key)))
ValueError: the 'axes' parameter is not supported in the pandas implementation of transpose()
```

(`ser.interpolate` doesn't raise, just forgets to interpolate)

**Comment by jreback** [comment]
i am suggesting a refactor completely of interpolate actually

is very hacky now - this is just making it worse

to be honest i would prob not touch this now

many many other fish to fry

**Comment by jbrockmendel** [comment]
OK.  I'll put a pin in this until I can come up with something more elegant.  Mind keeping it open for now so it doesn't fall off my radar screen?

**Pull request closed without merging**